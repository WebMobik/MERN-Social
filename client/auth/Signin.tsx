import React, { ChangeEvent, MouseEventHandler, useState } from 'react'
import { Redirect } from 'react-router-dom'
import auth from './auth-helper'
import { signin } from '../api/auth'
import {
  Avatar,
  Button,
  Card,
  CardContent,
  TextField,
  Typography,
} from '@material-ui/core'
import LockOutlinedIcon from '@material-ui/icons/LockOutlined';
import useStyles from '../styles/stylesForm'

const SignIn: React.FC = () => {
  const styles = useStyles()
  const [values, setValues] = useState({
    email: '',
    password: '',
    error: null,
    redirectToRefresh: false,
  })

  const handlerSubmit: MouseEventHandler<HTMLFormElement> = (e) => {
    e.preventDefault()
    const user = {
      email: values.email || undefined,
      password: values.password || undefined,
    }
    
    signin(user).then((data) => {
      if (data.error) {
        setValues({ ...values, error: data.error })
      } else {
        auth.authenticate(data, () => {
          setValues({ ...values, error: '', redirectToRefresh: true })
        })
      }
    })
  }

  const handlerChange = (name: string) => (event: ChangeEvent<{ value: string }>) => {
    setValues({ ...values, [name]: event.target.value })
  }

  return !values.redirectToRefresh ? (
    <Card className={styles.card}>
      <CardContent>
        <Avatar className={styles.avatar}>
          <LockOutlinedIcon />
        </Avatar>
        <Typography component="h1" variant="h5" className={styles.titleText}>
          Sign in
        </Typography>
        <form className={styles.form} onSubmit={handlerSubmit} >
          <TextField
            onChange={handlerChange('email')}
            value={values.email}
            variant="outlined"
            margin="normal"
            required
            fullWidth
            id="email"
            label="Email Address"
            name="email"
            autoComplete="email"
            autoFocus
          />
          <TextField
            onChange={handlerChange('password')}
            value={values.password}
            variant="outlined"
            margin="normal"
            required
            fullWidth
            name="password"
            label="Password"
            type="password"
            id="password"
            autoComplete="current-password"
          />
          <Button
            type="submit"
            fullWidth
            variant="contained"
            color="primary"
            className={styles.submit}
          >
            Sign In
          </Button>
        </form>
      </CardContent>
    </Card>
  ) : (
    <Redirect to="/" />
  )
}

export default SignIn
