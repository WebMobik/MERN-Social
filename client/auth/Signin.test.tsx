import React from 'react'
import {fireEvent, render, screen, waitFor} from '@testing-library/react'
import {act} from 'react-dom/test-utils'
import '@testing-library/jest-dom';
import Signin from './Signin'

const mockThen = jest.fn()

jest.mock('./api-auth', () => ({
    signin: () => ({
        then: mockThen
    })
}))

jest.mock('./auth-helper', () => ({
    authenticate: jest.fn()
}))

const TEST_INPUT_DATA = {
    EMAIL: 'test@mail.ru',
    PASSWORD: 'test123',
}

describe('<Signin /> test', () => {
    const Component = () => render(<Signin />)

    it('should be mount snap', () => {
        expect(Component()).toMatchSnapshot()
    })

    it('input form data', async () => {
        Component()
        const EmailInput = screen.getByPlaceholderText('Email')
        const PasswordInput = screen.getByPlaceholderText('Password')
        act(() => {
            fireEvent.change(EmailInput, { target: { value: TEST_INPUT_DATA.EMAIL } })
        })
        act(() => {
            fireEvent.change(PasswordInput, { target: { value: TEST_INPUT_DATA.PASSWORD } })
        })
        expect(screen.getByDisplayValue(TEST_INPUT_DATA.EMAIL)).toBeTruthy()
        expect(screen.getByDisplayValue(TEST_INPUT_DATA.PASSWORD)).toBeTruthy()

        act(() => {
            fireEvent.click(screen.getByText('Submit'))
        })

        await waitFor(() => expect(mockThen).toBeCalledTimes(1))
    })
})
