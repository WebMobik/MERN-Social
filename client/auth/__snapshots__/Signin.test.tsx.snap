// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signin /> test should be mount snap 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiCard-root makeStyles-card-1 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiCardContent-root"
        >
          <p
            class="MuiTypography-root makeStyles-titleText-3 MuiTypography-body1"
          >
            Sign In
          </p>
          <form
            class="makeStyles-form-2"
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                data-shrink="false"
                for="email"
                id="email-label"
              >
                Email
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input"
                  id="email"
                  placeholder="Email"
                  type="email"
                  value=""
                />
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                data-shrink="false"
                for="password"
                id="password-label"
              >
                Password
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input"
                  id="password"
                  placeholder="Password"
                  type="password"
                  value=""
                />
              </div>
            </div>
            <div
              class="MuiCardActions-root MuiCardActions-spacing"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-btnSuccess-4 MuiButton-containedPrimary"
                tabindex="0"
                type="submit"
              >
                <span
                  class="MuiButton-label"
                >
                  Submit
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiCard-root makeStyles-card-1 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <p
          class="MuiTypography-root makeStyles-titleText-3 MuiTypography-body1"
        >
          Sign In
        </p>
        <form
          class="makeStyles-form-2"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
              for="email"
              id="email-label"
            >
              Email
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="email"
                placeholder="Email"
                type="email"
                value=""
              />
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
              for="password"
              id="password-label"
            >
              Password
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="password"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-btnSuccess-4 MuiButton-containedPrimary"
              tabindex="0"
              type="submit"
            >
              <span
                class="MuiButton-label"
              >
                Submit
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
